src/Features/CSharp/Portable/AddBraces/CSharpAddBracesDiagnosticAnalyzer.cs

```diff
@@ -86,5 +108,172 @@ public void AnalyzeNode(SyntaxNodeAnalysisContext context)
+                return false;
+            }
+
+            // Check the part of the statement preceding the embedded statement
+            if (!FormattingRangeHelper.AreTwoTokensOnSameLine(embeddedStatement.GetFirstToken(), embeddedStatement.GetLastToken()))
+            {
+                // The embedded statement does not fit on one line. Examples:
+                //
+                //   if (something)
+                //     obj.Method(   // <-- This embedded statement spans two lines.
+                //       arg);
+                return true;
+            }
+
+            // Check the embedded statement itself
+            var lastTokenBeforeEmbeddedStatement = embeddedStatement.GetFirstToken().GetPreviousToken();
+            if (!FormattingRangeHelper.AreTwoTokensOnSameLine(statement.GetFirstToken(), lastTokenBeforeEmbeddedStatement))
+            {
+                // The part of the statement preceding the embedded statement does not fit on one line. Examples:
+                //
+                //   for (int i = 0; // <-- The initializer/condition/increment are on separate lines
+                //        i < 10;
+                //        i++)
+                //     SomeMethod();
+                return true;
+            }
+
+            // Check the part of the statement following the embedded statement, but only if it exists
+            if (statement.GetLastToken() != embeddedStatement.GetLastToken())
+            {
+                if (statement is IfStatementSyntax ifStatement && ifStatement.Statement == embeddedStatement)
+                {
+                    // The embedded statement is followed by an 'else' clause, which may span multiple lines without
+                    // triggering a braces requirement, such as this:
+                    //
+                    //   if (true)
+                    //     return;
+                    //   else          // <-- this else clause is two lines, but is not considered a multiline context
+                    //     return;
+                    //
+                    // ---
+                    // INTENTIONAL FALLTHROUGH
+                }
+                else
+                {
+                    var firstTokenAfterEmbeddedStatement = embeddedStatement.GetLastToken().GetNextToken();
+                    if (!FormattingRangeHelper.AreTwoTokensOnSameLine(firstTokenAfterEmbeddedStatement, statement.GetLastToken()))
+                    {
+                        // The part of the statement following the embedded statement does not fit on one line. Examples:
+                        //
+                        //   do
+                        //     SomeMethod();
+                        //   while (x < 0 ||    // <-- This condition is split across multiple lines.
+                        //          x > 10);
+                        return true;
+                    }
+                }
+            }
+
+            return false;
```

Comment:
having read the comment, i'm confident this method is correct :)  I am also now ok with your use of `if (statement.GetLastToken() != embeddedStatement.GetLastToken())` as i can see that it's a generalized way of handling the third chunk of a potential statement.